### PR TITLE
task/SWTCH-1542 ipfs module

### PR DESCRIPTION
**Pull request by JGiter**
- Added new ipfs module, which responsible for:
  - resolving from cluster or through gateway, depending on whether the claim was pinned
  - pinning
  - authenticating access to ipfs

These changes are tested in local Switchboard. `IPFS_CLUSTER_ROOT_URL` was set to our develop cluster, `IPFS_CLIENT_URL` was set to https://ipfs.io/
passport-did-auth wasn't yet upgraded to use cluster. It should not be a problem, because passport-did-auth and ssi-hub use same ipfs network, just accessing it through different protocols

### CODE REVIEW

File: src/modules/ipfs/ipfs.service.ts [hunk 1]
```diff
@@ -26,4 +34,47 @@ export class IPFSService {
       return false;
     }
   }
+
+  /**
+   * Get claim from cluster. If claim isn't found tries to get from gateway
+   *
+   * @param cid Content identifier of the credntial persisted on IPFS
```
Comment (nichonien) [inline src/modules/ipfs/ipfs.service.ts:41]:
```suggestion
   * @param cid Content identifier of the credential persisted on IPFS
```

File: .env.dev [hunk 2]
```diff
@@ -41,13 +41,15 @@ ASSETS_SYNC_HISTORY_INTERVAL_IN_HOURS=21
...
 
 # IPFS
-IPFS_CLUSTER_ROOT=$IPFS_CLUSTER_ROOT
-IPFS_CLUSTER_USER=$IPFS_CLUSTER_USER
-IPFS_CLUSTER_PASSWORD=$IPFS_CLUSTER_PASSWORD
-IPFS_CLIENT_HOST=<PUT_YOUR_VALUE_HERE>
-IPFS_CLIENT_PORT=<PUT_YOUR_VALUE_HERE>
-IPFS_CLIENT_PROJECT_ID=<PUT_YOUR_VALUE_HERE>
-IPFS_CLIENT_PROJECT_SECRET=<PUT_YOUR_VALUE_HERE>
+IPFS_CLUSTER_ROOT # Cluster can be deployed locally with e2e/setupIpfsCluster/index.ts
```
Comment (nichonien) [inline .env.dev:44]:
```suggestion
IPFS_CLUSTER_ROOT_URL # Cluster can be deployed locally with e2e/setupIpfsCluster/index.ts
```
IMO this is more clear for someone who doesn't have seen the code / implementation.

**Review by nichonien — APPROVED** [review]
Great work @JGiter , thanks!

File: .gitignore [hunk 3]
```diff
@@ -2,7 +2,6 @@
 /dist
 /node_modules
 /src/ethers
-.env
```
Comment (JGiter) [inline .gitignore:5]: it is also ignored in line 39

File: e2e/app.e2e.spec.ts [hunk 4]
```diff
@@ -8,11 +8,13 @@ import { appConfig } from '../src/common/test.utils';
 import { authTestSuite } from './auth';
 import { claimTestSuite } from './claim';
 import { statusList2021TestSuite } from './status-list';
-import { shutDownIpfsDaemon, spawnIpfsDaemon } from './setup-ipfs';
+import { shutdownIpfsCluster, spawnIpfsCluster } from './setupIpfsCluster';
```
Comment (nichonien) [inline e2e/app.e2e.spec.ts:11]: @JGiter We are not bringing up `IPFSGateway` for test cases, just relying on `Cluster` tests?

**Comment by jrhender** [comment]
@JGiter Could you resolve the conflict in `did.processor.ts` so that this PR could be merged please?

**Pull request closed without merging**